text/3366-diagnostic-attribute-namespace.md

```diff
@@ -0,0 +1,232 @@
+```rust
+#[diagnostic::on_unimplemented(
+    if(Self = std::string::String),
+    note = "That's only emitted if Self == std::string::String"
+)]
+#[diagnostic::on_unimplemented(
+    if(A = String), // Refers to whatever `String` is in the current scope
+    note = "That's only emitted if A == String",
+)]
+#[diagnostic::on_unimplemented(
+    if(any(A = i32, Self = i32)),
+    note = "That's emitted if A or Self is a i32",
+)]
+// this attribute will not have any affect as
+// the attribute above will always match before
+#[diagnostic::on_unimplemented(
+    if(all(A = i32, Self = i32)),
+    note = "That's emitted if A and Self is a i32"
+)]
+#[diagnostic::on_unimplemented(
+    if(not(A = String)),
+    // and implicitly `A` is not a `i32` as that case is
+    // matched earlier
+    note = "That's emitted if A is not a `String`"
+)]
+#[diagnostic::on_unimplemented(
+    message="message",
+    label="label",
+    note="That's emitted if neither of the condition above are meet",
+)]
+trait MyIterator<A> {
+    fn next(&mut self) -> A;
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+A possible drawback is that this feature adds additional complexity to the compiler implementation. The compiler needs to handle an additional attribute namespace with at least one additional attribute. 
+
+Another drawback is that crates might hint lower quality error messages than the compiler itself. Technically the compiler would be free to ignore such hints, practically I would assume that it is impossible to judge the quality of such error messages in an automated way.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+This proposal tries to improve error messages generated by rustc. It would give crate authors a tool to influence what error message is emitted in a certain situation, as they might sometimes want to provide specific details on certain error conditions. Not implementing this proposal would result in the current status quo. Currently the compiler always shows a "general" error message, even if it would be helpful to show additional details.
+
+There are alternatives for the naming of the `#[diagnostic]` namespace:
+
+* Use a `#[rustc]` namespace for these attributes. This would signifies that these are rustc specific extensions. We likely want to encourage other rust implementations to utilise these information as well, therefore a more general attribute namespace seems to be a better solution.
+
+There are alternative designs for the proposed `on_unimplemented` attribute:
+
+* The `on()` based filtering might be replaceable by placing the attribute on negative trait impls. This would turn a filter like
```

Comment:
`on()` has not been defined in the above description of `on_unimplemented`? it seems like the above description changed that to `of`.

Reply:
Good catch. That should be address in https://github.com/rust-lang/rfcs/pull/3368/commits/cb8f2fc5aec3b4f2d35a6f017fd7e9e59eb91fa7